Heapster bound names

brianhuffman opened the pull request:
This PR reverts "bug fix: binders with the same string names were causing unintended shadowing" (2e3c63d37919939db204c52108ce7546866f14e9), which was already fixed in #12, and merges #12 into the `wip-heapster` branch.

brianhuffman committed:
Translate bound names so as to avoid shadowing or using reserved keywords.

brianhuffman committed:
Revert "bug fix: binders with the same string names were causing unintended shadowing"

This reverts commit 2e3c63d37919939db204c52108ce7546866f14e9.

The same bug was already fixed in PR #12.

eddywestbrook commented:
@brianhuffman So, I like how this change handles reserved words in Coq, and the more generalized `nextVariant` approach for generating fresh names.

Unfortunately, the version in this PR still has some weird alpha-equivalence bug (or bugs?). I just tried it out with Heapster, which, because it auto-generates SAW core, re-uses the same names a lot and creates a lot of name shadowing.

The only way I was able to fix alpha-equivalence bugs in this code was to get rid of the `go env t` pattern, which was problematic because it was duplicating the environment information in two different places, the state monad and also the `env` argument to `go`. I found this really confusing and hard to get right. It should either use only the monad, or it should get rid of the monad and only use the `env` argument to `go` (which is effectively a `Reader` monad anyway).

My approach is still not ideal, though, since using side effects / mutable state to talk about name-binding contexts is certainly confusing. For instance, having `translateParams` translate the names of binders and also modify the local context so they are considered bound after the call is a little hard to keep in mind. Instead, this code should ideally use a `Reader` monad, where the variable context can only be changed in a locally-scoped way for particular sub-computations. It could then have a function like

```
withBoundNames :: [String] -> ([Coq.Ident] -> m a) -> m a
```

that would bind fresh Coq identifiers for each of the supplied `String` argument names, pass these Coq identifiers to the supplied function, and run its returned computation in the variable context extended with those identifiers. The benefit would be that it would be really clear when the variable context is being extended and what the duration of that extension is.

Anyway, not sure where to go from here...

brianhuffman commented:
I agree that it would probably be nicer to use a proper reader/state monad instead of just using state for everything. The code would probably be easier to understand and to maintain that way.

brianhuffman commented:
Ok, I'm pretty sure I know what's wrong. The issue is that in #12 I forgot to update the `translatePi` function to remap bound variable names. It should work if we add another small patch to do that. I probably missed that in my limited testing because I never translated any SAW terms that used dependent types.

@eddywestbrook did you notice any problems with lambda-bound variables, or just with Pi-bound ones? If it's only on Pi, then I think we can fix this pretty easily.

There are some other changes that I'd like to see, though, to reduce tech debt. Forgetting to translate bound variable names on a Pi wouldn't have been possible if saw-core and Coq-AST identifiers were not both the same type `String`; I think it would be a good idea to redefine type `Ident` from `Language.Coq.AST` as a `newtype` wrapper instead of just a type synonym for `String`. Also, we should do the reader-monad refactoring you suggested, and use bracketing combinators like `withBoundNames` instead of the `modify $ over localEnvironment (n :)` pattern that we have now. Probably @benjaminselfridge is the right person to implement those changes.

brianhuffman committed:
Fix `translatePi` so that it properly translates bound variables.

This patch fixes an oversight in the #12 PR. Previously it
incorrectly failed to use the name-collision-avoiding
`translateLocalIdent` function, and it similarly failed to
update the `localEnvironment` with the bound name.

brianhuffman commented:
@eddywestbrook I just added a new commit aba230b68bf06700e845d00f452e5c7230342483 to the PR, which should (hopefully) fix the remaining alpha-equivalence problems you were seeing. Check out a fresh copy of the `heapster-bound-names` branch and let me know if it works now.

brianhuffman committed:
Fix incorrect local variable environment when translating lambdas.

This patch fixes another problem that was missed in PR #12.
Previously when translating the body of a lambda, the translator
would populate the local variable environment with the *old*
variable names instead of using the new ones. This was causing
incorrect output when a lambda variable had been renamed.

brianhuffman committed:
Fix incorrect local variable environment when translating terms.

This patch fixes a few more instances of redundant and/or wrong
manipulations of the local variable environment when translating
terms. Most of these changes were adapted from the wip-heapster
branch (revision 2e3c63d3).

brianhuffman committed:
Merge branch 'bound-names' into heapster-bound-names

eddywestbrook reviewed (approved):
This seems to work in our local tests
